### PR TITLE
simplify remat partial eval parameterization

### PR DESCRIPTION
**Pull request by mattjj**
The main win here is reducing the number of arguments for the function that parameterizes _remat_partial_eval (which exists so that _remat_patial_eval can used both with remat and invertible ad features). Also a bit more work is now done in the caller. I think it's a net reduction in lines, though the diff --stat doesn't show it because I wrapped some long lines and added some comments.

I also included a fix to _remat_partial_eval that is needed in #3370, though I don't think it's needed on master. It was easier to include the fix now, and means #3370 can be just a touch smaller.

Both these changes made rebasing #3370 easier!

I moved the remat tests into their own class in api_test.py, but otherwise that file didn't change.

### CODE REVIEW

File: jax/interpreters/partial_eval.py
Comment (apaszke) [inline jax/interpreters/partial_eval.py:712]: What's the benefit of doing this so early? It seems like it complicates the `partial_eval_jaxpr` part a little, because we also have to pass in consts as arguments.

File: jax/interpreters/invertible_ad.py
Comment (apaszke) [inline jax/interpreters/invertible_ad.py:57]: I guess this could have been `trace.new_const(t.pval.get_known())`?

**Review by apaszke — APPROVED** [review]
Nice, thanks for simplifying that!

File: jax/interpreters/partial_eval.py
Comment (mattjj) [reply jax/interpreters/partial_eval.py:712]: It's kind of specific to #3370, where the consts that come out might be tracers. Then it's easier just to treat them like tracers from the start, i.e. basically treat them like the env values.

Maybe we can iteratively simplify in follow-ups if we find ways to improve? My preference would be to iterate after landing #3370 though :)

File: jax/interpreters/invertible_ad.py
Comment (mattjj) [reply jax/interpreters/invertible_ad.py:57]: I'll give that a shot!

File: jax/interpreters/partial_eval.py
Comment (apaszke) [reply jax/interpreters/partial_eval.py:712]: Sure, I don't mean to block this PR! It just seemed a little surprising.

File: jax/interpreters/invertible_ad.py
Comment (mattjj) [reply jax/interpreters/invertible_ad.py]: Worked like a charm!

**Comment by mattjj** [comment]
Internal tests pass!

**Comment by mattjj** [comment]
Oops, forgot to switch to "squash" again... Somehow that seems to keep getting un-selected as my default.

**Pull request merged**
